spacy/ml/tb_framework.pyx

```diff
@@ -133,27 +145,36 @@ def init(
 
 
 def forward(model, docs_moves: Tuple[List[Doc], TransitionSystem], is_train: bool):
-    nF = model.get_dim("nF")
-    tok2vec = model.get_ref("tok2vec")
+    beam_width = model.attrs.get("beam_width", 1)
     lower_pad = model.get_param("lower_pad")
-    lower_W = model.get_param("lower_W")
+    tok2vec = model.get_ref("tok2vec")
+
+    docs, moves = docs_moves
+    states = moves.init_batch(docs)
+    tokvecs, backprop_tok2vec = tok2vec(docs, is_train)
+    tokvecs = model.ops.xp.vstack((tokvecs, lower_pad))
+    feats, backprop_feats = _forward_precomputable_affine(model, tokvecs, is_train)
+    seen_mask = _get_seen_mask(model)
+
+    if beam_width == 1 and not is_train and isinstance(model.ops, NumpyOps):
+        return forward_cpu(model, moves, states, feats, seen_mask)
+    else:
+        return forward_thinc(model, moves, states, tokvecs, backprop_tok2vec, feats, backprop_feats, seen_mask, is_train)
+
+
+def forward_thinc(model: Model, moves: TransitionSystem, states: List[StateClass], tokvecs, backprop_tok2vec, feats, backprop_feats, seen_mask, is_train: bool):
```

Comment:
I'd love to come up with a better name for this function, `forward_thinc` sounds a bit odd to me, but I don't have a good suggestion right now :|

Reply:
Yeah, the names are not good. `forward_cpu` is also odd, since `forward_thinc` can also be on the CPU. Maybe:

* `_forward_greedy_cpu`
* `_forward`

?

Reply:
I think that's probably as good as we can do. I'll change it to your suggested names.